File: backend/app/services/recode_ranges.py
"""Range bands for a recode definition — #823(d), 2026-08-31.

Banding a continuous variable ("18–29 → Young") without typing one row per
distinct value. The authoring cost this removes is real and was measured twice:
**72 rows** for GSS `age`, **39 rows on a 48-row dataset** (#830(j)) — so it is
not a large-data problem, and *Category Group*, the rule type literally named for
grouping, had the identical shape.

## Where this sits in the match rule

A recode has always matched a cell by case-insensitive equality against
``mapping``'s keys. A range adds a SECOND channel, and the order is fixed:

1. **the null set** — declared-missing, recognized-N/A, or the definition's own
   ``exclude_values``. Unchanged, and it still runs FIRST (J-D1 / #594): a
   sentinel like ``-99`` on `age` must stay NULL even when a band covers it
   numerically. This is the reason the range channel could not simply be
   "another mapping".
2. **an explicit ``mapping`` key** — the specific beats the general, so a
   researcher who bands `0–120` and then maps `"99" → Refused` gets what they
   wrote.
3. **the first matching range**, in declared order.

⚠️ **`resolve_range_output` is the ONE implementation of step 3**, consumed by
both `compute_value` (per cell) and `plan_definition_over_column` (per distinct
value). #542b is the standing record of what a second copy costs here: the two
paths once disagreed about a stray non-numeric mapping value, so one cell got two
different numbers depending on which path computed it. The client mirrors this in
`lib/recode-ranges.ts` and is deliberately NOT authoritative.

## Parsing a cell

Through ``dataset_import._strip_numeric`` — THE cell-to-number rule — never a
local ``float()``. It is what makes a ``"1,200"`` parse the same way it does
everywhere else, and it already rejects the non-finite values #689 depends on.
A cell that does not parse simply does not match a range; it is not an error.
"""

import json
import logging

logger = logging.getLogger(__name__)

#: A band list is short by nature — a banding with more entries than this is
#: almost certainly a paste, and the per-value `mapping` is the right shape for
#: a long enumeration. Mirrors `MAX_MISSING_RULES`' reasoning, one concept over.
MAX_RECODE_RANGES = 50


class RangeBandError(ValueError):
    """A malformed band list. Raised at the WRITE path, never at apply time.

    ⚠️ Deliberately a `ValueError` subclass so the router's existing
    `except ValueError` arms keep working, and deliberately NOT raised from the
    read/apply path: `apply_definition_to_column` is on the STARTUP path via
    `repair_reverse_recode_mappings` (#794), so a raise there fires during boot
    on existing data. The read side degrades to "no bands" and logs.
    """


def _coerce_bound(raw, field: str):
    """A bound is a finite number or ``None`` (open end)."""
    if raw is None or raw == "":
        return None
    try:
        value = float(raw)
    except (TypeError, ValueError):
        raise RangeBandError(f"Range {field} must be a number.")
    # `math.isfinite` rather than a bare comparison: a bare `Infinity` is not
    # JSON-compliant and would 500 the response that tried to return it (#689,
    # and the `.sav` ±inf trap in the missing-values importer).
    if value != value or value in (float("inf"), float("-inf")):
        raise RangeBandError(f"Range {field} must be a finite number.")
    return value


def normalize_ranges(raw, *, allow_output_text: bool = True) -> list[dict]:
    """Validate a band list on the way IN. Raises `RangeBandError`.

    Mirrors `services/missing_values._validate_rule`'s shape rules — at least one
    bound, bounds finite, ``lo <= hi`` — and adds the one that differs: the
    OUTPUT is required, because here it is the rule's result rather than display
    metadata.

    ``allow_output_text`` is False for a SCALE_MAP, whose output must be a
    number: a scale map writes `value_numeric`, and a band emitting a string
    there would land in `unmapped` at apply time with no explanation the
    researcher could act on.
    """
    if raw is None:
        return []
    if not isinstance(raw, list):
        raise RangeBandError("Ranges must be a list.")
    if len(raw) > MAX_RECODE_RANGES:
        raise RangeBandError(
            f"At most {MAX_RECODE_RANGES} ranges can be declared (received {len(raw)})."
        )

    out: list[dict] = []
    for entry in raw:
        if not isinstance(entry, dict):
            raise RangeBandError("Each range must be an object.")
        lo = _coerce_bound(entry.get("lo"), "low bound")
        hi = _coerce_bound(entry.get("hi"), "high bound")
        if lo is None and hi is None:
            raise RangeBandError("A range needs a low or a high value.")
        if lo is not None and hi is not None and lo > hi:
            raise RangeBandError(f"Range {lo} to {hi} is backwards.")

        if "output" not in entry or entry["output"] is None or entry["output"] == "":
            raise RangeBandError("Every range needs a value to map to.")
        output = entry["output"]
        if isinstance(output, str) and not allow_output_text:
            try:
                output = float(output)
            except ValueError:
                raise RangeBandError(
                    f'A scale map\'s ranges must map to numbers — "{output}" is text. '
                    "Use a category group to band into named groups."
                )
        elif not isinstance(output, (str, int, float)):
            raise RangeBandError("A range's value must be a number or a name.")

        out.append({"lo": lo, "hi": hi, "output": output})
    return out


def parse_ranges(raw_json: str | None) -> list[dict]:
    """Read a stored band list. Never raises — the read path must not.

    ⚠️ Shape is re-checked on the way OUT, not only on the way in. A stored
    non-list (legacy data, a hand-edited database, a future field written by a
    newer build) would otherwise reach the matcher and raise mid-apply, on the
    STARTUP path. Degrading to "no bands" logs loudly and leaves the mapping
    channel working.
    """
    if not raw_json:
        return []
    try:
        parsed = json.loads(raw_json)
    except (json.JSONDecodeError, TypeError):
        logger.warning("Unparseable recode ranges JSON; ignoring: %r", raw_json[:120])
        return []
    if not isinstance(parsed, list):
        logger.warning("Recode ranges is not a list; ignoring: %r", parsed)
        return []

    out: list[dict] = []
    for entry in parsed:
        if not isinstance(entry, dict) or "output" not in entry:
            continue
        lo, hi = entry.get("lo"), entry.get("hi")
        if lo is not None and not isinstance(lo, (int, float)):
            continue
        if hi is not None and not isinstance(hi, (int, float)):
            continue
        if lo is None and hi is None:
            continue
        out.append({"lo": lo, "hi": hi, "output": entry["output"]})
    return out


def resolve_range_output(value_text: str, ranges: list[dict]):
    """The band a cell falls in, or ``None``.

    🔴 **THE range match, with exactly one implementation.** Both backend
    matchers call it — `compute_value` per cell and `plan_definition_over_column`
    per distinct value — because a cell must get the same number whichever path
    computed it (#542b).

    Bounds are INCLUSIVE at both ends, which is what a researcher writing
    "18 to 29" means. Adjacent bands are therefore the author's responsibility
    (18–29, 30–44), and overlap is resolved by ORDER: the first declared band
    that contains the value wins, so a narrow special case can be listed above a
    catch-all.
    """
    if not ranges or value_text is None:
        return None
    # Imported here rather than at module scope only to keep this module free of
    # the import-time cost of the adapter package; `dataset_import` does not
    # import this service, so a module-level import would also be safe.
    from .dataset_import import _strip_numeric

    number = _strip_numeric(value_text.strip())
    if number is None:
        return None
    for band in ranges:
        lo, hi = band.get("lo"), band.get("hi")
        if lo is not None and number < lo:
            continue
        if hi is not None and number > hi:
            continue
        return band["output"]
    return None

File: frontend/src/lib/recode-ranges.ts
/**
 * Range bands, client side — #823(d), 2026-08-31.
 *
 * The MIRROR of `services/recode_ranges.py`, and deliberately not the authority.
 * The server decides what `value_numeric` holds; this exists so the Data view's
 * display lens and the rule editor's preview show the same answer the apply will
 * produce.
 *
 * 🔴 **Why a mirror is required rather than optional here.** `EditableCell`
 * resolves a cell through the active definition entirely client-side. Without a
 * band channel it would render a banded cell as unmapped plain text while the
 * server has written the band's code into `value_numeric` — the #578
 * display-vs-storage drift, on the one grid a researcher checks a recode
 * against.
 *
 * ⚠️ **What this mirror deliberately CANNOT do: the null set.** Deciding whether
 * a cell is missing needs `_is_na`'s prefix list and the column's declaration,
 * which is exactly the derivation #600 forbids the client from attempting. So
 * the caller applies its own exclude check first (as `EditableCell` already
 * does) and this answers only "which band, if any". A cell the server NULLs and
 * this bands is not a disagreement about bands — it is the caller skipping the
 * check that runs before them.
 */

/** One band. `lo`/`hi` are inclusive; either may be null for an open end. */
export interface RecodeRange {
  lo: number | null
  hi: number | null
  output: number | string
}

/**
 * The glyphs `_strip_numeric` removes, in its order and its set:
 * `$ € £ ¥ , %`. Removed ANYWHERE in the string, not just at the ends — so
 * `"1$2"` is 12 on both sides. That is odd, and it is the rule; a mirror that
 * "improved" on it would disagree with the number actually stored.
 */
const STRIPPED_GLYPHS = /[$€£¥,%]/g

/**
 * What Python's `float()` accepts, minus what it doesn't: a plain decimal
 * literal with an optional sign and exponent.
 *
 * ⚠️ **The gate exists to keep the client from being WIDER than the server.**
 * `Number()` accepts radix literals — `0x10` is 16, `0b101` is 5, `0o17` is 15 —
 * and `float()` rejects all three. Client-wider is the dangerous direction: the
 * grid would show a band the apply will never write.
 */
const DECIMAL_LITERAL = /^[+-]?(?:\d+(?:\.\d*)?|\.\d+)(?:[eE][+-]?\d+)?$/

/**
 * Parse a cell the way the backend's `_strip_numeric` does, so a `"$50,000"`
 * lands in the same band on both sides.
 *
 * 🔴 **#862 — this docstring used to make that claim while stripping commas and
 * nothing else, and `Number()` is not `float()`.** MEASURED across 39 inputs,
 * **16 disagreed**: every currency and percent cell (`"$50,000"` → 50000 on the
 * server, nothing on the client — on exactly the column types bands exist for),
 * every radix literal in the opposite direction, and `"1_000"`. The
 * cross-language fixture carried no glyph-bearing cell, so both suites passed.
 *
 * The rule, in the backend's order: strip the ends · remove the glyphs anywhere
 * · strip again · parse as a decimal literal.
 *
 * ⚠️ **`%` does NOT divide by 100.** `_strip_numeric("45%")` is 45, so a band of
 * 40–50 claims it. Surprising, deliberate, and the mirror's job is to agree.
 *
 * ⚠️ **Underscores between digits are removed, because `float()` accepts them**
 * (PEP 515) and `Number()` does not. The lookbehind/lookahead pair is what
 * matches Python's *"single underscores, between digits only"*: `"1_0"` → 10,
 * while `"1__0"`, `"_10"` and `"10_"` keep theirs and fail the grammar, exactly
 * as `float()` refuses them.
 *
 * 🔴 **THE ONE REMAINING DIVERGENCE, STATED RATHER THAN DISCOVERED LATER:
 * non-ASCII decimal digits.** `float()` accepts them (measured: `"٣٤"` → 34,
 * `"１２３"` → 123, `"۵"` → 5, via CPython's decimal-to-ASCII transform); this
 * does not, so such a cell renders as raw text while the server bands it. That
 * is the SAFE direction — under-claiming, not showing a code the apply will not
 * write — and closing it needs a Unicode `Nd`-block transform, which is more
 * cleverness than the case has earned.
 *
 * ⚠️ **`NFKC` normalisation is NOT the cheap fix for that, and was rejected on a
 * measurement:** it would newly accept `"²"` as 2 and `"５％"` as 5, both of
 * which `float()` refuses — i.e. it buys one narrow case by opening two in the
 * dangerous direction.
 */
export function parseCellNumber(text: string | null | undefined): number | null {
  if (text === null || text === undefined) return null
  const trimmed = String(text).trim()
  if (trimmed === '') return null
  const cleaned = trimmed.replace(STRIPPED_GLYPHS, '').trim()
  const withoutSeparators = cleaned.replace(/(?<=\d)_(?=\d)/g, '')
  if (!DECIMAL_LITERAL.test(withoutSeparators)) return null
  const n = Number(withoutSeparators)
  return Number.isFinite(n) ? n : null
}

/**
 * The band a cell falls in, or `undefined` when none does.
 *
 * `undefined` rather than `null` so a caller can distinguish "no band matched"
 * from a band whose output legitimately IS null-ish — and so it composes with
 * `mapping`'s lookup, which also yields `undefined` on a miss.
 *
 * Bounds are INCLUSIVE at both ends and the FIRST matching band wins, so a
 * narrow special case listed above a catch-all shadows it. Both rules mirror
 * `resolve_range_output`.
 */
export function resolveRangeOutput(
  valueText: string | null | undefined,
  ranges: RecodeRange[] | undefined | null,
): number | string | undefined {
  if (!ranges || ranges.length === 0) return undefined
  const n = parseCellNumber(valueText)
  if (n === null) return undefined
  for (const band of ranges) {
    if (band.lo !== null && band.lo !== undefined && n < band.lo) continue
    if (band.hi !== null && band.hi !== undefined && n > band.hi) continue
    return band.output
  }
  return undefined
}

/** Human wording for one band — used by the editor and by the cell tooltip. */
export function describeRange(band: RecodeRange): string {
  const { lo, hi } = band
  if (lo !== null && hi !== null) return lo === hi ? `${lo}` : `${lo} to ${hi}`
  if (lo !== null) return `${lo} and above`
  if (hi !== null) return `${hi} and below`
  return 'any value'
}

/**
 * Index pairs whose bands overlap — for DISCLOSURE, never refusal.
 *
 * Overlap is legal and sometimes deliberate, so the editor SAYS which row
 * shadows which rather than blocking the save.
 *
 * ⚠️ **The CLIENT owns this rule outright (#866).** There was a Python twin,
 * `recode_ranges.describe_overlaps`, with no production caller and nothing
 * binding it to this function — the server never refuses an overlap and never
 * reports one, because the disclosure is an AUTHORING concern. It was deleted
 * rather than pinned: an unconsumed mirror can only drift.
 */
export function rangeOverlaps(ranges: RecodeRange[]): [number, number][] {
  const pairs: [number, number][] = []
  for (let i = 0; i < ranges.length; i++) {
    for (let j = i + 1; j < ranges.length; j++) {
      const aLo = ranges[i].lo ?? -Infinity
      const aHi = ranges[i].hi ?? Infinity
      const bLo = ranges[j].lo ?? -Infinity
      const bHi = ranges[j].hi ?? Infinity
      if (aLo <= bHi && bLo <= aHi) pairs.push([i, j])
    }
  }
  return pairs
}

/** A row mid-edit: bounds stay strings so the inputs stay controlled. */
export interface RangeRow {
  lo: string
  hi: string
  output: string
}

export interface RangeValidation {
  ok: boolean
  msg: string
  ranges: RecodeRange[]
  badRow?: number
}

const isBlankRow = (r: RangeRow) =>
  r.lo.trim() === '' && r.hi.trim() === '' && r.output.trim() === ''

/**
 * One editor row's bounds, or `null` when the row cannot take part in a
 * comparison at all.
 *
 * 🔴 **ONE eligibility rule, and #863 is what two of them cost.** The editor
 * used to filter with *"at least one bound is finite"* and then hand the
 * survivors to `rangeOverlaps`, which compares with `lo ?? -Infinity` — and
 * `??` does **not** catch `NaN`, so a row like `{lo: "abc", hi: "40"}` passed
 * the filter and then compared false against everything. Two rules for one
 * question, disagreeing silently.
 *
 * A row takes part iff **at least one bound is written AND every bound that is
 * written parses to a finite number.** A half-typed `"abc"` is not an open end:
 * treating it as one would invent a band the researcher never wrote and flag an
 * overlap against it.
 */
function rowBounds(row: RangeRow): RecodeRange | null {
  const parse = (raw: string): number | null | undefined => {
    const t = raw.trim()
    if (t === '') return null
    const n = Number(t)
    return Number.isFinite(n) ? n : undefined
  }
  const lo = parse(row.lo)
  const hi = parse(row.hi)
  if (lo === undefined || hi === undefined) return null
  if (lo === null && hi === null) return null
  return { lo, hi, output: row.output }
}

/**
 * Which EDITOR ROWS overlap — as indices into `rows`, not into some subset.
 *
 * 🔴 **#863: the notice named the wrong rows, and named blank ones.** The editor
 * filtered incomplete rows out and then printed `rangeOverlaps`' indices as row
 * numbers — but those index the FILTERED list, so every skipped row above a pair
 * shifted the answer down. Reproduced live: rows `[blank, 18–29, 25–40]`
 * reported *"Range 1 and range 2 overlap"*, naming an empty row and neither of
 * the two that actually did.
 *
 * ⚠️ **The index mapping lives here, once, next to the eligibility rule** —
 * putting it back at the call site is what made a pure, correct function give a
 * wrong answer. `rangeOverlaps` stays exactly as it was: the pure pair-wise
 * comparison, which the client owns outright (#866).
 *
 * ⚠️ Deliberately **not** memoised. It is O(n²) over a list capped at
 * `MAX_RECODE_RANGES` (50), i.e. ≤1,225 numeric comparisons on the extreme
 * input, against a `useMemo` whose dependency array is one more thing to get
 * wrong on a component that re-renders per keystroke anyway.
 */
export function rowOverlaps(rows: RangeRow[]): [number, number][] {
  const eligible: { band: RecodeRange; index: number }[] = []
  rows.forEach((row, index) => {
    const band = rowBounds(row)
    if (band) eligible.push({ band, index })
  })
  return rangeOverlaps(eligible.map(e => e.band)).map(
    ([a, b]) => [eligible[a].index, eligible[b].index] as [number, number],
  )
}

/**
 * How the editor names one row: its ordinal, plus its bounds when it has any.
 *
 * ⚠️ **The ordinal is the half that always works.** `describeRange` alone gives
 * two identical bands two identical names — the *"N buttons called Remove"*
 * problem (#785) this file's own docstring claims to avoid — and gives a blank
 * row the actively wrong *"any value"*, which is what an unbounded band means.
 * The ordinal also matches what the inputs are called (`Range 3 from`), so a
 * reader hears one identifier for one row.
 */
export function describeRow(rows: RangeRow[], index: number): string {
  const band = rowBounds(rows[index])
  return band ? `range ${index + 1} (${describeRange(band)})` : `range ${index + 1}`
}

/**
 * Validate the rows and build the wire payload — the mirror of
 * `normalize_ranges`, with the SAME refusals in the same order so a researcher
 * meets the message here rather than as a 400.
 *
 * `numericOutput` is true for a scale map, whose bands must produce numbers.
 * ⚠️ Blank rows are skipped rather than refused: a trailing empty row is how an
 * editor looks mid-edit, not an error to explain.
 */
export function buildRangePayload(
  rows: RangeRow[],
  numericOutput: boolean,
): RangeValidation {
  const ranges: RecodeRange[] = []
  for (let i = 0; i < rows.length; i++) {
    const r = rows[i]
    if (isBlankRow(r)) continue
    const loRaw = r.lo.trim()
    const hiRaw = r.hi.trim()
    const lo = loRaw === '' ? null : Number(loRaw)
    const hi = hiRaw === '' ? null : Number(hiRaw)
    if (lo === null && hi === null) {
      return { ok: false, msg: 'A range needs a low or a high value.', ranges: [], badRow: i }
    }
    if ((lo !== null && !Number.isFinite(lo)) || (hi !== null && !Number.isFinite(hi))) {
      return { ok: false, msg: 'Range bounds must be numbers.', ranges: [], badRow: i }
    }
    if (lo !== null && hi !== null && lo > hi) {
      return { ok: false, msg: `Range ${lo} to ${hi} is backwards.`, ranges: [], badRow: i }
    }
    const output = r.output.trim()
    if (output === '') {
      return { ok: false, msg: 'Every range needs a value to map to.', ranges: [], badRow: i }
    }
    if (numericOutput && !Number.isFinite(Number(output))) {
      return {
        ok: false,
        msg: `A scale map's ranges must map to numbers — "${output}" is text.`,
        ranges: [],
        badRow: i,
      }
    }
    ranges.push({ lo, hi, output: numericOutput ? Number(output) : output })
  }
  return { ok: true, msg: '', ranges }
}

/** Seed editor rows from a stored band list. */
export function rangesToRows(ranges: RecodeRange[] | undefined | null): RangeRow[] {
  if (!ranges) return []
  return ranges.map(r => ({
    lo: r.lo === null || r.lo === undefined ? '' : String(r.lo),
    hi: r.hi === null || r.hi === undefined ? '' : String(r.hi),
    output: String(r.output),
  }))
}
